networkx/classes/backends.py

```diff
@@ -0,0 +1,227 @@
+    _registered_algorithms[name] = wrapped_func
+
+
+def _dispatch(func=None, *, name=None):
+    """Dispatches to a backend algorithm
+    when the first argument is a backend graph-like object.
+    """
+    # Allow any of the following decorator forms:
+    #  - @_dispatch
+    #  - @_dispatch()
+    #  - @_dispatch("override_name")
+    #  - @_dispatch(name="override_name")
+    if func is None:
+        if name is None:
+            return _dispatch
+        return functools.partial(_dispatch, name=name)
+    if isinstance(func, str):
+        return functools.partial(_dispatch, name=func)
+    # If name not provided, use the name of the function
+    if name is None:
+        name = func.__name__
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwds):
+        graph = args[0]
+        if hasattr(graph, "__networkx_plugin__") and plugins:
+            plugin_name = graph.__networkx_plugin__
+            if plugin_name in plugins:
+                backend = plugins[plugin_name].load()
+                if hasattr(backend, name):
+                    return getattr(backend, name).__call__(*args, **kwds)
+                else:
+                    raise NetworkXNotImplemented(
+                        f"'{name}' not implemented by {plugin_name}"
+                    )
+        return func(*args, **kwds)
+
+    _register_algo(name, wrapper)
+    return wrapper
+
+
+def test_override_dispatch(func=None, *, name=None):
+    """Auto-converts the first argument into the backend equivalent,
+    causing the dispatching mechanism to trigger for every
+    decorated algorithm."""
+    if func is None:
+        if name is None:
+            return test_override_dispatch
+        return functools.partial(test_override_dispatch, name=name)
+    if isinstance(func, str):
+        return functools.partial(test_override_dispatch, name=func)
+    # If name not provided, use the name of the function
+    if name is None:
+        name = func.__name__
+
+    sig = inspect.signature(func)
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwds):
+        backend = plugins[plugin_name].load()
+        if not hasattr(backend, name):
+            pytest.xfail(f"'{name}' not implemented by {plugin_name}")
+        bound = sig.bind(*args, **kwds)
+        bound.apply_defaults()
+        graph, *args = args
+        # Convert graph into backend graph-like object
+        #   Include the weight label, if provided to the algorithm
+        weight = None
+        if "weight" in bound.arguments:
+            weight = bound.arguments["weight"]
+        elif "data" in bound.arguments and "default" in bound.arguments:
+            # This case exists for several MultiGraph edge algorithms
+            if isinstance(bound.arguments["data"], str):
+                weight = bound.arguments["data"]
+            elif bound.arguments["data"]:
+                weight = "weight"
+        graph = backend.convert(graph, weight=weight)
```

Comment:
These hard-coded attributes seem fine for now (especially `weight`, since that is pretty universal and tends to have the same meaning across all functions). In the future, it'd be good to have a more flexible mechanism (or a more formalized standard for argument pass through) in place.

One potential issue - what about the case where `weight` is a Callable? Is this something backends can (should? must?) be expected to support?